Sources/ArgumentParser/Parsable Types/ParsableArgumentsValidation.swift

```diff
@@ -0,0 +1,86 @@
+//===----------------------------------------------------------*- swift -*-===//
+//
+// This source file is part of the Swift Argument Parser open source project
+//
+// Copyright (c) 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+extension ParsableArguments {
+  static func _validate() throws {
+    try ParsableArgumentsCodingKeyValidator.validate(self)
+  }
+}
+
+/// Ensure that all arguments have corresponding coding keys
+struct ParsableArgumentsCodingKeyValidator {
+  
+  private struct Validator: Decoder {
+    let argumentKeys: [String]
+    
+    enum ValidationResult: Swift.Error {
+      case success
+      case missingCodingKeys([String])
+    }
+    
+    let codingPath: [CodingKey] = []
+    let userInfo: [CodingUserInfoKey : Any] = [:]
+    
+    func unkeyedContainer() throws -> UnkeyedDecodingContainer {
+      fatalError()
+    }
+    
+    func singleValueContainer() throws -> SingleValueDecodingContainer {
+      fatalError()
+    }
+    
+    func container<Key>(keyedBy type: Key.Type) throws -> KeyedDecodingContainer<Key> where Key : CodingKey {
+      let missingKeys = argumentKeys.filter { Key(stringValue: $0) == nil }
+      if missingKeys.isEmpty {
+        throw ValidationResult.success
+      } else {
+        throw ValidationResult.missingCodingKeys(missingKeys)
+      }
+    }
+  }
+  
+  /// This error indicates that an option, a flag, or an argument of
+  /// a `ParsableArguments` is defined without a corresponding `CodingKey`.
+  struct Error: Swift.Error {
+    let parsableArgumentsType: ParsableArguments.Type
+    let missingCodingKeys: [String]
+  }
+  
+  static func validate(_ type: ParsableArguments.Type) throws {
+    let argumentKeys: [String] = Mirror(reflecting: type.init())
```

Comment:
This looks like a generally useful pattern that we’ll want for a variety of validations. What do you think about breaking this out as a `ParsableArguments` extension?

Reply:
I prefer making this a protocol. So we can have multiple validators, separate the logics, and we can write tests for each of them.

```Swift
protocol ParsableArgumentsValidator {
  static func validate(_ type: ParsableArguments.Type) throws
}

// for hasPositionalFollowingRepeated
struct ArgumentPositionValidator: ParsableArgumentsValidator {}

// for coding keys
struct ParsableArgumentsCodingKeyValidator: ParsableArgumentsValidator {}

extension ParsableArguments {
  static func _validate() throws {
    let validators: [ParsableArgumentsValidator.Type] = [
      ArgumentPositionValidator.self,
      ParsableArgumentsCodingKeyValidator.self
    ]
    for validator in validators {
      try validator.validate(self)
    }
  }
}
```